set_variant(): try next requested variant if one fails to import

merlinND opened the pull request:
## Description

If the user specifies several variants in order, e.g.:

```py
mi.set_variant("cuda_ad_rgb", "llvm_ad_rgb")
```

and they are all compiled (available) but some fail to import, then keep trying with the next requested variant instead of throwing an exception.

This could happen e.g. when requested a CUDA variant with the NVIDIA driver installed / CUDA available but no GPU available, or when LLVM is not installed.

## Testing

On a machine with the NVIDIA driver installed,

```bash
CUDA_VISIBLE_DEVICES= python -c 'import mitsuba as mi; mi.set_variant("cuda_ad_rgb", "llvm_ad_rgb"); print(mi.variant())'
```

Old behavior: raises `ImportError` and exits.
New behavior: fails to load `cuda_ad_rgb`, but correctly goes to the second choice `llvm_ad_rgb`.

The old behavior can still be obtained by requesting a single variant at a time (no fallback): `mi.set_variant("cuda_ad_rgb")`

## Checklist

<!-- Please make sure to complete this checklist before requesting a review. -->

- [x] My code follows the [style guidelines](https://mitsuba.readthedocs.io/en/latest/src/developer_guide.html#coding-style) of this project
- [x] My changes generate no new warnings
- [x] My code also compiles for `cuda_*` and `llvm_*` variants. If you can't test this, please leave below
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I cleaned the commit history and removed any "Merge" commits
- [x] I give permission that the Mitsuba 3 project may redistribute my contributions under the terms of its [license](https://github.com/mitsuba-renderer/mitsuba/blob/master/LICENSE)

merlinND commented on src/python/alias.cpp:
```diff
@@ -116,8 +116,35 @@ static void set_variant(nb::args args) {
+            // We only allow such failures as long as we have more variants to try.
+            if (!is_last && e.matches(PyExc_ImportError)) {
+                // nb::str var_str = requested_variant;
+                // jit_log(LogLevel::Debug,
```
I would have liked to make a debug log here, but `jit_log` is not linked against this module. Is there an easy way to log here?

njroussel replied:
I see two options:
* We link `libmitsuba.so` to `mitsuba_alias.so` so that we can call `mitsuba::Thread::thread()->logger()->log(level, nullptr, __FILE__, __LINE__, ...)`
* We're guaranteed to have loaded `mitsuba_ext.so`, so we theoretically do a Python call to `mitsuba.Log(Level, message)` through Python.  

I'd personally prefer the second one.

merlinND replied:
Thanks, doing a Log call through Python works :+1:

merlinND commented on src/python/alias.cpp:
```diff
@@ -95,16 +95,16 @@ static nb::object variant_module(nb::handle variant) {
-            new_variant = nb::borrow(arg);
-            break;
+            // Variant is at least compiled, we can attempt to use it.
+            valid_variants.append(nb::borrow(arg));
```
I'm not certain about the ref counting here.

njroussel replied:
Fairly certain that the `borrow` is ok. Are there currently any leaks?
I can imagine that a failed `set_variant` itself might leak some stuff too actually.

wjakob replied:
You should not need the borrow at all IMO.

njroussel reviewed (commented):
I like this! I've had it on my backlog to have `mi.variants()` only report valid variants, but there's one nasty problem: if that logic is ever wrong/broken, we'll be skipping hundreds of tests on the CI. Keeping this fallback logic in `set_variant()` shouldn't ever mess with how we've setup the test suite :+1:

merlinND commented:
Thanks for the review @njroussel and @wjakob!
I've rebased and incorporated the feedback.

Currently, both on `master` and with this PR, failing to set a variant results in a lot of leaks when exiting, even if the exception is caught and we then successfully set another variant:

```py
# CUDA_VISIBLE_DEVICES= python ./test_failed_variant.py

import mitsuba as mi

if __name__ == "__main__":
    mi.set_log_level(mi.LogLevel.Debug)
    try:
        mi.set_variant("cuda_ad_rgb")
    except ImportError as e:
        print("Failed to set variant")

    print("Current variant:", mi.variant())
    # Still leaks if we then set a valid variant
    mi.set_variant("llvm_ad_rgb")
    print("Current variant:", mi.variant())
```

```
nanobind: leaked 23 instances!
 - leaked instance 0x769448183008 of type "mitsuba.Color0f"
 - leaked instance 0x769443f3bac8 of type "drjit.cuda.ad.TensorXf"
 - leaked instance 0x769443f1ebd0 of type "drjit.scalar.Array3f"
 - ... skipped remainder
nanobind: leaked 273 types!
 - leaked type "drjit.cuda.ad.Int"
 - leaked type "mitsuba.Ray3d"
 - ... skipped remainder
nanobind: leaked 1486 functions!
- ...
```

it's unrelated to this PR, though.

merlinND commented:
Pinging this PR, would it be okay to merge?

The latest commit ([4306f55](https://github.com/mitsuba-renderer/mitsuba3/pull/1522/commits/4306f55688b1f0deef8828eb8dc15e6600ba1468)) avoids partial initialization of variants when a backend turns out to be unavailable, which used to result in leaks (even before this PR).

njroussel commented:
LGTM!

Thanks for looking at the leaks too :)
